Add Typescript annotations to NumberInput

jpsorensen opened the pull request:
Closes #12548 

Convert existing propType definitions to Typescript annotations on the NumberInput component. This is part of a broader effort to add Typescript annotations to components, tracked in #12513.

#### Changelog

**Changed**

- NumberInput.js has been converted to tsx
- useNormalizedInputProps jsdocs updated to reflect more accurate input types, since Typescript will infer a Typescript definition from those docs
- added jpsorensen to contributors list

#### Testing / Reviewing

- In the packages/react directory, run yarn build, yarn storybook. In the storybook, navigate to NumberInput and confirm that the component's behavior remains unchanged.
- In NumberInput.stories.js and NumberInput-test.js,  add `// @ts-check` as the first line in the file (temporarily, for the sake of this test). Confirm that the files do not have compilation errors. Make changes to the components in those files - change number to string, etc - and confirm that compilation errors appear.

jpsorensen committed:
refactor(NumberInputSkeleton): add typescript annotations

Convert existing propType definitions to Typescript annotations on the
NumberInputSkeleton component. This is part of a broader effort to add
Typescript annotations to components, tracked in #12513. Closes #12549.

Type annotation changes only; no breaking feature changes.

jpsorensen committed:
refactor(NumberInput): add typescript annotations

Convert existing propType definitions to Typescript annotations on the
NumberInput component. This is part of a broader effort to add
Typescript annotations to components, tracked in #12513. Closes #12548.

Type annotation changes only; no breaking feature changes.

jpsorensen committed:
Merge branch 'main' into refactor/12548-number-input-types

jpsorensen committed:
refactor(NumberInput): reactNodeLike to ReactNode

Change ReactNodeLike (which is a PropType type) to ReactNode.
Clean up console log.

Type annotation changes only; no breaking changes.

jpsorensen committed:
fix(NumberInput): helperText optional

Made helperText property optional in NumberInput Typescript annotations.

jpsorensen committed:
Merge branch 'main' into refactor/12548-number-input-types

tay1orjones commented on packages/react/src/components/NumberInput/NumberInput.tsx:
```diff
@@ -70,7 +238,9 @@ const NumberInput = React.forwardRef(function NumberInput(props, forwardRef) {
     useState(controlledValue);
   const inputRef = useRef(null);
-  const ref = useMergedRefs([forwardRef, inputRef]);
+  const ref = useMergedRefs([forwardRef, inputRef]) as
+    | LegacyRef<HTMLInputElement>
+    | undefined;
```
@mbarrer is this what you were thinking for the forwardRef pattern?

tay1orjones commented on packages/react/src/components/NumberInput/NumberInput.tsx:
```diff
@@ -110,11 +280,16 @@ const NumberInput = React.forwardRef(function NumberInput(props, forwardRef) {
+    // uncontrolled component (https://reactjs.org/link/controlled-components).
+    // This condition can only be reached if controlledValue is not equal to prevControlledValue.
+    // controlledValue can only be undefined AND not equal to prevControlledValue if the above
+    // rule is violated (the component changed from controlled to uncontrolled).
```
This is good context! I don't think it needs to live in the source though
```suggestion

```

tay1orjones commented on packages/react/src/components/NumberInput/NumberInput.tsx:
```diff
@@ -110,11 +280,16 @@ const NumberInput = React.forwardRef(function NumberInput(props, forwardRef) {
+    // This condition can only be reached if controlledValue is not equal to prevControlledValue.
+    // controlledValue can only be undefined AND not equal to prevControlledValue if the above
+    // rule is violated (the component changed from controlled to uncontrolled).
+    setValue(controlledValue!);
```
I do think an eslint disable line comment will need to be added here. It can be scoped to the individual rule

tay1orjones commented on packages/react/src/components/NumberInput/NumberInput.tsx:
```diff
@@ -138,27 +313,34 @@ const NumberInput = React.forwardRef(function NumberInput(props, forwardRef) {
 
   const outerElementClasses = cx(`${prefix}--form-item`, {
-    [customClassName]: !!customClassName,
+    [customClassName!]: !!customClassName,
```
Same thing here with the eslint disable line rule. It can be scoped to the specific rule

jpsorensen replied:
Ah thanks for catching! I hadn't noticed that lint passes but with warnings. Now that I have ESLint set up in VSCode I'll catch these much earlier.

tay1orjones commented on packages/react/src/components/NumberInput/NumberInput.tsx:
```diff
@@ -138,27 +313,34 @@ const NumberInput = React.forwardRef(function NumberInput(props, forwardRef) {
+  // normalizedProps.icon actually has type CarbonIconType | undefined, from the DefinitelyTyped
+  // definition for icons-react. We use any here instead to avoid adding a dependency to that
+  // library. If those types become available without that dependency we should update the type
+  // of normalizedProps.icon.
```
Again, good context but I think we can leave it out of the source
```suggestion

```

tay1orjones commented on packages/react/src/components/NumberInput/NumberInput.tsx:
```diff
@@ -217,7 +397,9 @@ const NumberInput = React.forwardRef(function NumberInput(props, forwardRef) {
                   const state = {
-                    value: clamp(max, min, parseInt(value) - step),
+                    // parseInt will coerce its argument to a string, so even if value is a number
+                    // we will be safe - hence the typecast to convince Typescript.
```
```suggestion
```

tay1orjones commented on packages/react/src/components/NumberInput/NumberInput.tsx:
```diff
@@ -242,7 +424,9 @@ const NumberInput = React.forwardRef(function NumberInput(props, forwardRef) {
                   const state = {
-                    value: clamp(max, min, parseInt(value) + step),
+                    // parseInt will coerce its argument to a string, so even if value is a number
+                    // we will be safe - hence the typecast to convince Typescript.
```
```suggestion
```

tay1orjones reviewed (changes requested):
This is great! 🎉 Just a few pieces to address below. In the future instead of making typing-related comments in the source I think it might be better to just annotate the diff in the PR with a comment.

Any idea why the syntax highlighting in the diff isn't working? I'm not seeing any of it in the .tsx file diff. Not a huge deal, just odd.

jpsorensen committed:
Update packages/react/src/components/NumberInput/NumberInput.tsx

Remove verbose comments

Co-authored-by: Taylor Jones <tay1orjones@users.noreply.github.com>

jpsorensen committed:
Apply suggestions from code review

Remove extra comments

Co-authored-by: Taylor Jones <tay1orjones@users.noreply.github.com>

jpsorensen committed:
fix(lint): lint failures in NumberInput

Disable non-null-assertion checks on two specific lines where we
know the assertion to be safe (one because react forbids components
from changing from controlled to not controlled, and the other because
the value might be undefined but the function still behaves as intended).

andreancardona reviewed (approved):
looks good to me. Sorry for the delay on the review, thank you for the contribution! 🎉

jpsorensen committed:
Merge branch 'main' into refactor/12548-number-input-types

tay1orjones reviewed (approved):
@jpsorensen Thanks I appreciate it! This looks good 👍

andreancardona committed:
Merge branch 'main' into refactor/12548-number-input-types

kodiakhq[bot] committed:
Merge branch 'main' into refactor/12548-number-input-types
